src/util/live_performance.js

```diff
@@ -0,0 +1,137 @@
+// @flow
+
+import window from './window.js';
+import {version as sdkVersion} from '../../package.json';
+import type Projection from '../geo/projection/projection.js';
+import {
+    isMapboxHTTPStyleURL,
+    isMapboxHTTPTileJSONURL,
+    isMapboxHTTPSpriteURL,
+    isMapboxHTTPFontsURL
+} from './mapbox.js';
+
+type LivePerformanceMetrics = {
+    counters: Array<Object>,
+    metadata: Array<Object>,
+    attributes: Array<Object>
+};
+
+export type LivePerformanceData = {
+    interactionRange: [number, number],
+    width: number,
+    height: number,
+    terrainEnabled: boolean,
+    fogEnabled: boolean,
+    projection: Projection,
+    renderer: ?string,
+    vendor: ?string
+};
+
+function categorize(arr, fn) {
+    const obj = {};
+    if (arr) {
+        for (const item of arr) {
+            const category = fn(item);
+            if (obj[category] === undefined) {
+                obj[category] = [];
+            }
+            obj[category].push(item);
+        }
+    }
+    return obj;
+}
+
+function getTransferRangePerResourceType(resourceTimers) {
```

Comment:
Do you think ranges make sense for things like tilejson? Will the range be a meaningful value if we can't tell how many requests are in that range?

Reply:
It's true that the "Range" is a bit of an opaque concept, and doesn't say whether it's collecting one or several resources, it's more of a grouping for a resource category type.

An option would be to add specific resource transfer times as such:
```
- sprite0TransferStart
- sprite0TransferEnd
- sprite2TransferStart
- sprite2TransferEnd
- font0TransferStart
- font1TransferStart
- ...
```
but that makes it more tedious on the other end to interpret the collected data. It might be ok be ok with to be more opaque for an easier manipulation of the data, even if that means loosing some granularity.

>Will the range be a meaningful value if we can't tell how many requests are in that range?

It would fairly to add a counter for each of these category in this PR, e.g.:

```
- spriteFetchCount: 3
- fontRangeFetchCount 5
...
```

Another useful information is the number of cached resources for that category, but that's something that we'd include separately from this PR (Refer https://mapbox.atlassian.net/browse/GLJS-220).

Reply:
I'm not sure. Do what seems best to you and we can always revise!

Reply:
Added number of requests in https://github.com/mapbox/mapbox-gl-js/pull/12343/commits/9b80532eeb00ccefbb4b7b7d9b9ca496cf40aab4.